transport/socket_transport.go

```diff
@@ -0,0 +1,503 @@
+		log.Errorf("failure in client connection readLoop: %v", r)
+		if err := s.conn.Close(); err != nil {
+			// Ignore
+		}
+	}
+}
+
+func (s *SocketConnection) responseHandler(buff []byte) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	version := SocketTransportVersion(binary.BigEndian.Uint16(buff))
+	if version != SocketTransportV1 {
+		return errors.Errorf("invalid transport version: %d - only version %d supported", version, SocketTransportV1)
+	}
+	correlationID := int64(binary.BigEndian.Uint64(buff[2:]))
+	ch, ok := s.responseChannels[correlationID]
+	if !ok {
+		return errors.Errorf("no client response handler found with id %d", correlationID)
+	}
+	delete(s.responseChannels, correlationID)
+	// Next byte signifies whether error or not
+	if buff[10] == 1 {
+		// Error
+		errorCode := binary.BigEndian.Uint16(buff[11:])
+		msgLen := binary.BigEndian.Uint32(buff[13:])
+		msg := string(buff[17 : 17+msgLen])
+		ch <- responseHolder{
+			err: common.NewTektiteError(common.ErrCode(errorCode), msg),
+		}
+	} else {
+		// Must copy as connection reader re-uses the buffer
+		copied := common.ByteSliceCopy(buff[11:])
+		ch <- responseHolder{
+			response: copied,
+		}
+	}
+	return nil
+}
+
+func (s *SocketConnection) SendRPC(handlerID int, request []byte) ([]byte, error) {
+	ch, err := s.sendRequest(handlerID, request)
+	if err != nil {
+		// If we get an error writing the request we consider the connection broken, so we close it
+		// The user can then retry creating a new connection and sending the RPC again, potentially after a delay.
+		if err := s.Close(); err != nil {
+			// Ignore
+		}
+		return nil, err
+	}
+	holder := <-ch
+	return holder.response, holder.err
+}
+
+func (s *SocketConnection) sendRequest(handlerID int, request []byte) (chan responseHolder, error) {
+	buff, ch := s.createRequest(handlerID, request)
+	// Set a write deadline so the write doesn't block for a long time in case the other side of the TCP connection
+	// disappears
+	if err := s.conn.SetWriteDeadline(time.Now().Add(writeTimeout)); err != nil {
+		return nil, err
+	}
+	_, err := s.conn.Write(buff)
+	if err != nil {
+		return nil, convertNetworkError(err)
+	}
+	return ch, nil
+}
+
+func convertNetworkError(err error) error {
+	// We convert to Tektite unavailable errors, as they are retryable
+	return common.NewTektiteErrorf(common.Unavailable, "transport error when sending rpc: %v", err)
+}
+
+func (s *SocketConnection) createRequest(handlerID int, request []byte) ([]byte, chan responseHolder) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	length := len(request) + 22
+	buff := make([]byte, length)
+	binary.BigEndian.PutUint32(buff, uint32(length-4))
+	binary.BigEndian.PutUint16(buff[4:], uint16(SocketTransportV1))
+	binary.BigEndian.PutUint64(buff[6:], uint64(s.correlationIDSequence))
+	binary.BigEndian.PutUint64(buff[14:], uint64(handlerID))
+	copy(buff[22:], request)
+	ch := make(chan responseHolder, 1)
+	s.responseChannels[s.correlationIDSequence] = ch
```

Comment:
is this perf sensitive?
wondering if we just need to lock these two lines if so

Reply:
On second thought it needs wider lock as it accesses s.correlationIDSequence earlier